feat(output)!: result aggregation as a task

maugde opened the pull request:
Linked to [ANT-3118](https://gopro-tickets.rte-france.com/browse/ANT-3118)

The purpose of this PR is to compute aggregation operations in task processes. In that way, heavy operations won't block main process. 
First, the user asks for the generation of a file that should contain results of an aggregation operation. 
Then, the user can ask for these results when the task is completed.


BREAKING CHANGE: this PR introduces a new endpoint `v1/downloads/{download_id}/metadata` that retrieves metadata about the state of a download.

It changes return values of several output endpoints: 
- `v1/studies/{uuid}/outputs/{output_id}/aggregate/areas/mc-all`
- `v1/studies/{uuid}/outputs/{output_id}/aggregate/links/mc-ind`
- `v1/studies/{uuid}/outputs/{output_id}/aggregate/links/mc-all`
- `v1/studies/{uuid}/outputs/{output_id}/aggregate/areas/mc-ind`

And their respectives redirecting versions:
- `/studies/{uuid}/areas/aggregate/mc-all/{output_id}`
- `/studies/{uuid}/links/aggregate/mc-ind/{output_id}`
- `/studies/{uuid}/links/aggregate/mc-all/{output_id}`
- `/studies/{uuid}/areas/aggregate/mc-ind/{output_id}`

Commit:
feat(output_service): add AggregateOutputTask and its type, update output endpoints

Commit:
feat(output_service): add endpoint and update the task process

replace task method, add endpoint, exception `TaskResultNotReady`, model and migration file

Commit:
test(output_service): add and update tests, rename variables for clarity, and update comments

MartinBelthle commented on antarest/study/storage/output_service.py:
```diff
@@ -484,18 +490,95 @@ def aggregate_output_data(
+
+        task_id = self._task_service.add_task(
+            aggregate_output_task,
+            "Aggregate output",
```
The msg could be a bit clearer with study id and output id

MartinBelthle commented on antarest/study/storage/output_service.py:
```diff
@@ -484,18 +490,95 @@ def aggregate_output_data(
+                output_aggregation = db.session.get(OutputAggregation, task_id)
+            download_file = self._file_transfer_manager.fetch_download(output_aggregation.download_id)
+            return FileResponse(path=download_file.path, status_code=200)
+        elif task.status == TaskStatus.RUNNING:
```
We should replace this with in {TaskStatus.PENDING, TaskStatus.RUNNING}

MartinBelthle commented on antarest/study/storage/output_service.py:
```diff
@@ -484,18 +490,95 @@ def aggregate_output_data(
+        elif task.status == TaskStatus.RUNNING:
+            msg = task.result.message if task.result else ""
+            raise AggregatedOutputNotReady(task_id, msg)
+        else:
```
Here we're in the case of a failed task so I don't think the exception you introduced quite fit

MartinBelthle commented on antarest/study/storage/output_service.py:
```diff
@@ -484,18 +490,95 @@ def aggregate_output_data(
+            return FileResponse(path=download_file.path, status_code=200)
+        elif task.status == TaskStatus.RUNNING:
+            msg = task.result.message if task.result else ""
+            raise AggregatedOutputNotReady(task_id, msg)
```
Here I don't think we can have a task.result as the task is still running, i think you can remove this variable `msg` and simplify your exception

MartinBelthle commented on antarest/study/storage/output_service.py:
```diff
@@ -484,18 +490,95 @@ def aggregate_output_data(
+            msg = task.result.message if task.result else ""
+            raise AggregatedOutputNotReady(task_id, msg)
+        else:
+            msg = task.result.message if task.result else ""
```
Same remark here, I think you'll always have a result as the task ended

MartinBelthle commented on antarest/study/storage/output_service.py:
```diff
@@ -484,18 +490,95 @@ def aggregate_output_data(
+
+        if task.status == TaskStatus.COMPLETED:
+            with db():
+                output_aggregation = db.session.get(OutputAggregation, task_id)
```
Minor: In the code we almost always use query instead of get
ex: `db.session.query(FileDownload).get(download_id)`

MartinBelthle commented on tests/integration/raw_studies_blueprint/test_aggregate_raw_data.py:
```diff
@@ -455,6 +456,12 @@ def test_area_aggregation(
             res = client.get(f"/v1/studies/{internal_study_id}/areas/aggregate/mc-ind/{output_id}", params=params)
             assert res.status_code == 200, res.json()
+            task_id = res.json()
+            time.sleep(1.0)
```
Instead of `time.sleep` we should use `client.get(f"/v1/tasks/{task_id}?wait_for_completion=True")` to ensure our tests always pass

MartinBelthle reviewed (changes requested):
It's going in the right direction :). Unfortunately, you'll probably have several conflicts with Sylvain PR https://github.com/AntaresSimulatorTeam/AntaREST/pull/2484

Commit:
chore(output_service): minor fixes of tests, exceptions, and logs

MartinBelthle commented on antarest/core/exceptions.py:
```diff
@@ -336,6 +336,21 @@ def __init__(self) -> None:
+    def __init__(self, task_id: str) -> None:
+        super().__init__(
+            HTTPStatus.UNPROCESSABLE_ENTITY,
+            f"Cannot retrieve results for task '{task_id}'. Task results are not ready yet",
```
Really minor: Usually we do msg = ... and then super().init with the msg to be a bit clearer. As you did in the next exception

MartinBelthle commented on antarest/core/exceptions.py:
```diff
@@ -336,6 +336,21 @@ def __init__(self) -> None:
+
+
+class AggregatedOutputNotProcessed(HTTPException):
+    def __init__(self, task_id: str, info: Optional[str]) -> None:
```
I think here you can type info as a str to remove the if and this would lead to also putting a type: ignore when raising the Exception or an assert or a or "". I think the last solution is preferable

MartinBelthle commented on antarest/core/exceptions.py:
```diff
@@ -336,6 +336,21 @@ def __init__(self) -> None:
+        )
+
+
+class AggregatedOutputNotProcessed(HTTPException):
```
The exception name seems a bit confusing. I understand it as it's not done yet (like the other exception). I think simply putting `OutputAggregationFailed` would be more understandable

MartinBelthle commented on tests/integration/raw_studies_blueprint/test_aggregate_raw_data.py:
```diff
@@ -806,21 +876,31 @@ def test_aggregation_errors(
+            "The output_id that wasn't found should be in the message"
+        )
+
+        # Asserts that a 404 error is raised when the `economy/mc-all` folder does not exist
```
Minor: It's no longer a 404 but a 422 so you can change the msg

MartinBelthle reviewed (changes requested):
Minor comments. We'll have to take another look after resolving conflicts with Sylvain's PR (to be merged soon)

Commit:
chore(output_service): minor fixes on exceptions

Commit:
chore(output_service): use tasks to compute aggregation results

Commit:
feat(file_transfer): add endpoint, remove database models and version, remove exceptions

Commit:
test(aggregation): fix aggregation_raw_data tests

Commit:
chore(output): minor fix

MartinBelthle commented on antarest/core/filetransfer/web.py:
```diff
@@ -45,4 +51,57 @@ def fetch_download(download_id: str) -> Any:
+    @bp.get(
+        "/downloads/{download_id}/metadata",
+        tags=[APITag.downloads],
+        summary="Retrieve download file or information about the preparation of the download file",  # FIXME
```
there's still a fixme

MartinBelthle commented on antarest/core/filetransfer/web.py:
```diff
@@ -45,4 +51,57 @@ def fetch_download(download_id: str) -> Any:
+            if wait_for_availability:
+                try:
+                    task_service.await_task(task_id=sanitized_task_id, timeout_sec=timeout)
+                    task = task_service.status_task(sanitized_task_id)  # update task status
```
I think you should put this line line 85 to be clearer as for now my IDE says this line is useless

MartinBelthle commented on tests/integration/raw_studies_blueprint/test_aggregate_raw_data.py:
```diff
@@ -701,6 +794,15 @@ def test_different_formats(
             output_id = params.pop("output_id")
             res = client.get(f"/v1/studies/{internal_study_id}/links/aggregate/mc-all/{output_id}", params=params)
             assert res.status_code == 200, res.json()
+            file_download_task = res.json()
```
This whole process is really repetitive. Could you put this inside a function to simplify the test readability ?

MartinBelthle commented on tests/integration/raw_studies_blueprint/test_aggregate_raw_data.py:
```diff
@@ -881,9 +1037,88 @@ def test_columns_formatting(
+            f"v1/downloads/{fake_file_data_id}/metadata",
+            params={"task_id": fake_task_id, "wait_for_availability": True},
+        )
+        assert res.status_code == 404, res.json()["exception"] == "HTTPException"
```
You could add checks on the exception message. This goes for the rest of the test

MartinBelthle commented on tests/integration/raw_studies_blueprint/test_aggregate_raw_data.py:
```diff
@@ -881,9 +1037,88 @@ def test_columns_formatting(
+        res = client.get(
+            f"v1/downloads/{file_data_id}/metadata", params={"task_id": task_id, "wait_for_availability": False}
+        )
+        assert res.status_code == 202, res.json()["description"] == "The requested file is still in process."
```
I think we shouldn't use 202. Even if the description fits, it implies the request succeeded and so it will create confusion and the R scripts might fail unexpectedly (same for Antares-craft).

I don't see a nice Exception to use but by default we could use the 417

MartinBelthle reviewed (changes requested):
Seems okay, few comments

Commit:
chore(filetransfer): change return code, improve test and other minor fixes

MartinBelthle commented on antarest/core/filetransfer/web.py:
```diff
@@ -45,4 +51,58 @@ def fetch_download(download_id: str) -> Any:
+        task_id: str,
+        download_id: str,
+        wait_for_availability: bool = False,
+        timeout: int = DEFAULT_AWAIT_MAX_TIMEOUT,
```
It's weird to have a `timeout` here that you only use when `wait_for_availability` is True.
I think you can hard code it and remove it from the endpoint arguments

MartinBelthle commented on antarest/core/filetransfer/web.py:
```diff
@@ -45,4 +51,58 @@ def fetch_download(download_id: str) -> Any:
+                    task_service.await_task(task_id=sanitized_task_id, timeout_sec=timeout)
+                except concurrent.futures.TimeoutError:
+                    raise HTTPException(
+                        status_code=http.HTTPStatus.EXPECTATION_FAILED,
```
Here we should use HTTP 408 I think.
It's what we currently do inside the GET "/tasks/{task_id}" endpoint

MartinBelthle commented on antarest/core/filetransfer/web.py:
```diff
@@ -45,4 +51,58 @@ def fetch_download(download_id: str) -> Any:
+        sanitized_task_id = sanitize_uuid(task_id)
+        task = task_service.status_task(sanitized_task_id)
+
+        # the user ask for a timeout
```
This comment is not really clear and should be put after `if wait_for_availability`.
Could be renamed `The user wants to wait for the download to be available`

MartinBelthle commented on antarest/core/filetransfer/web.py:
```diff
@@ -45,4 +51,58 @@ def fetch_download(download_id: str) -> Any:
+                        status_code=http.HTTPStatus.EXPECTATION_FAILED,
+                        detail=f"The requested file is still in process after waiting for {timeout} seconds.",
+                    )
+            # the user did not ask for a timeout
```
Same here, the user does not really "ask" for a timeout

MartinBelthle commented on antarest/core/filetransfer/web.py:
```diff
@@ -45,4 +51,58 @@ def fetch_download(download_id: str) -> Any:
+                detail=f"The requested file was not successfully processed: {task.result.message}",  # type: ignore
+            )
+
+        try:
```
Don't think we need this try catch as we already have them inside the `fetch_download` method

MartinBelthle reviewed (changes requested):
Last round of comments for me

MartinBelthle commented:
You should mention in the PR comment that you're introducing a breaking change and also put a `!` in the PR title to show it

Commit:
chore(filetransfer): remove timeout argument and improve the fetch_download call

MartinBelthle reviewed (changes requested):
The more I look to the endpoints, the more I think the design is not what we wanted in the first place.

To me what we should be doing is:
- Endpoints `/aggregate`, just return `task_id` and `download_id` as the other info are always the same and useless
- Endpoint `/metadata`, return `FileDownloadTaskDTO`. If the user set `wait_...` check the status as you did already. If the user didn't just return the DTO as it is 

The user will then use the already existing /download endpoint to get the FileResponse

Commit:
refactor(outputs): refactor download and output endpoints

Commit:
test(outputs): update tests

MartinBelthle commented on antarest/core/filetransfer/web.py:
```diff
@@ -45,4 +45,22 @@ def fetch_download(download_id: str) -> Any:
+        sanitized_task_id = sanitize_uuid(task_id)
+        sanitized_download_id = sanitize_uuid(download_id)
+
+        # the task was successfully completed
```
This comment doesn't make sense anymore

MartinBelthle commented on antarest/study/storage/output_service.py:
```diff
@@ -467,10 +468,13 @@ def aggregate_output_data(
+        download_log: str,
         mc_years: Optional[Sequence[int]] = None,
-    ) -> Iterator[pd.DataFrame]:
+    ) -> Tuple[str, str]:
```
I think a class with 2 attributes: `task_id` and `download_id` would be better to maintain the code

MartinBelthle commented on antarest/study/web/output_blueprint.py:
```diff
@@ -203,7 +201,7 @@ def aggregate_areas_raw_data(
         columns_names: str = "",
         export_format: TableExportFormat = DEFAULT_EXPORT_FORMAT,
-    ) -> FileResponse:
+    ) -> Dict[str, str]:
```
Same remark here on the class

MartinBelthle commented on antarest/core/filetransfer/service.py:
```diff
@@ -207,3 +220,29 @@ def fetch_download(self, download_id: str) -> FileDownload:
+    def get_download_metadata(self, task_id: str, download_id: str, wait_for_availability: bool) -> FileDownloadTaskDTO:
+        task = self.task_service.status_task(task_id)
+
+        file_raw = self.repository.get(download_id)
```
`file_raw` could be renamed `download` or `file_download` for clarity sake

MartinBelthle commented on tests/integration/raw_studies_blueprint/test_aggregate_raw_data.py:
```diff
@@ -881,9 +1042,92 @@ def test_columns_formatting(
+        res = client.get(
+            f"v1/downloads/{download_id}/metadata", params={"task_id": task_id, "wait_for_availability": False}
+        )
+        assert res.status_code == 200, res.json()
```
Why don't we have an error here ? If we ask the output too early without waiting we should have an exception

MartinBelthle reviewed (changes requested):
Almost done I think

Commit:
chore(output): minor fixes

fix bad indentation in file transfer service, add class for typing

MartinBelthle commented on antarest/study/storage/output_service.py:
```diff
@@ -479,17 +483,71 @@ def aggregate_output_data(
             mc_years: list of monte-carlo years, if empty, all years are selected (only for mc-ind)
 
-        Returns: the aggregated data as a DataFrame
+        Returns: tuple of the file download id and the task id
```
Change the docstring accordingly

MartinBelthle commented on antarest/study/web/output_blueprint.py:
```diff
@@ -298,7 +296,7 @@ def aggregate_links_raw_data(
 
         Returns:
-            FileResponse that corresponds to a dataframe with the aggregated links raw data
+            id of the task that generates a dataframe with the aggregated links raw data  # FIXME
```
fixme comment

MartinBelthle commented on antarest/study/web/output_blueprint.py:
```diff
@@ -375,7 +371,7 @@ def aggregate_areas_raw_data__all(
 
         Returns:
-            FileResponse that corresponds to a dataframe with the aggregated areas raw data
+            id of the task that generates a dataframe with the aggregated areas raw data  # FIXME
```
same

MartinBelthle commented on antarest/study/web/output_blueprint.py:
```diff
@@ -449,7 +444,7 @@ def aggregate_links_raw_data__all(
 
         Returns:
-            FileResponse that corresponds to a dataframe with the aggregated links raw data
+            id of the task that generates a dataframe with the aggregated links raw data  # FIXME
```
same

MartinBelthle reviewed (changes requested):
Final comments on the doc

Commit:
chore(output): update comments

MartinBelthle reviewed (changes requested):
After discussion with Sylvain, I realized we don't need to return the `task_id` in the endpoint. We only need the `download_id` as a str. 
And instead of waiting the task, we can simply poll the download inside the db and wait for its attribute `ready` to be True

Sorry to change this but it would make way more sense

Commit:
chore(output): change return value of output aggregation endpoints

Commit:
fix(output): fix tests

MartinBelthle commented on antarest/core/filetransfer/service.py:
```diff
@@ -38,11 +48,13 @@ def __init__(
         self.config = config
         self.repository = repository
         self.event_bus = event_bus
+        self.task_service = task_service
```
Seems you don't need the task service anymore

MartinBelthle commented on antarest/core/filetransfer/service.py:
```diff
@@ -207,3 +219,30 @@ def fetch_download(self, download_id: str) -> FileDownload:
+            while time.time() < end and not download.ready and not download.failed:
+                with db():  # needs db context to refresh download
+                    download = self.repository.get(download_id)
+                    if not download:
```
Don't think you need this `if` as you checked this line 225

MartinBelthle commented on antarest/study/business/aggregator_management.py:
```diff
@@ -420,7 +420,7 @@ def aggregate_output_data(self) -> Iterator[pd.DataFrame]:
 
         logger.info(
-            f"Parsing {len(all_output_files)} {self.frequency.value} files"
+            f"Parsing {len(all_output_files)} {self.frequency.value} files "
```
Minor but you can remove this change I think

maugde replied:
I noticed that this log wasn't correctly displayed due to the missing space character ("filesto" instead of "files to"). Should I remove it anyway ?

MartinBelthle commented on tests/integration/raw_studies_blueprint/test_aggregate_raw_data.py:
```diff
@@ -434,6 +435,11 @@
 ]
 
 
+def get_file_and_task_ids(data: dict) -> Tuple[str, str]:
```
You can remove this function as it's not relevant anymore

Commit:
fix(output): minor fixes
remove task attribute to file transfer service, remove a typing condition and unused function

Commit:
fix(output): minor fixes
add typing hints

MartinBelthle reviewed (changes requested):
Tests are failing

Commit:
fix(tests): minor fixes

MartinBelthle commented on antarest/core/filetransfer/service.py:
```diff
@@ -207,3 +217,29 @@ def fetch_download(self, download_id: str) -> FileDownload:
+        if wait_for_availability:
+            end = time.time() + DEFAULT_AWAIT_MAX_TIMEOUT
+
+            # disable download variable typing since it will always be defined
```
Okay I didn't understand why you put the check, now I do. It would be better to remove the `type: ignore` and your comment and just do a `assert download is not None` line 234

After that I'll merge

Commit:
chore(file_transfer): add typing check test

MartinBelthle committed:
Merge branch 'dev' into feat/ANT-3118-result-aggregation-as-a-task
